state-chain/pallets/cf-broadcast/src/migrations/v3.rs

```diff
@@ -0,0 +1,283 @@
+		pub broadcast_id: BroadcastId,
+		pub attempt_count: AttemptCount,
+	}
+
+	#[derive(Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub struct BroadcastAttempt<T: Config<I>, I: 'static> {
+		pub broadcast_attempt_id: BroadcastAttemptId,
+		pub transaction_payload: TransactionFor<T, I>,
+		pub threshold_signature_payload: PayloadFor<T, I>,
+		pub transaction_out_id: TransactionOutIdFor<T, I>,
+	}
+
+	#[derive(Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub struct TransactionSigningAttempt<T: Config<I>, I: 'static> {
+		pub broadcast_attempt: BroadcastAttempt<T, I>,
+		pub nominee: T::ValidatorId,
+	}
+
+	#[frame_support::storage_alias]
+	pub type AwaitingBroadcast<T: Config<I>, I: 'static> =
+		StorageMap<Pallet<T, I>, Twox64Concat, BroadcastAttemptId, TransactionSigningAttempt<T, I>>;
+
+	#[frame_support::storage_alias]
+	pub type FailedBroadcasters<T: Config<I>, I: 'static> =
+		StorageMap<Pallet<T, I>, Twox64Concat, BroadcastId, Vec<<T as Chainflip>::ValidatorId>>;
+
+	#[frame_support::storage_alias]
+	pub type BroadcastRetryQueue<T: Config<I>, I: 'static> =
+		StorageValue<Pallet<T, I>, Vec<BroadcastAttempt<T, I>>>;
+}
+
+#[derive(Clone, Encode, Decode)]
+pub struct MigrationVerification<T: Chainflip> {
+	broadcasts_with_failed_broadcasters:
+		Vec<(BroadcastId, BTreeSet<<T as Chainflip>::ValidatorId>)>,
+	broadcast_retry_queue: Vec<BroadcastId>,
+}
+pub struct Migration<T: Config<I>, I: 'static>(PhantomData<(T, I)>);
+
+impl<T: Config<I>, I: 'static> OnRuntimeUpgrade for Migration<T, I> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// Awaiting broadcast: take the broadcast attempt with the highest attempt.
+		let mut latest_attempt = BTreeMap::new();
+		old::AwaitingBroadcast::<T, I>::iter_keys().for_each(
+			|old::BroadcastAttemptId { broadcast_id, attempt_count }| {
+				if match latest_attempt.get(&broadcast_id) {
+					Some(attempt) => attempt_count > *attempt,
+					None => true,
+				} {
+					latest_attempt.insert(broadcast_id, attempt_count);
+				}
+			},
+		);
+
+		// Migrate data from old storage into new storage.
+		latest_attempt.into_iter().for_each(|(broadcast_id, attempt_count)| {
+			if let Some(attempt) = old::AwaitingBroadcast::<T, I>::get(old::BroadcastAttemptId {
+				broadcast_id,
+				attempt_count,
+			}) {
+				AwaitingBroadcast::<T, I>::insert(
+					broadcast_id,
+					TransactionSigningAttempt {
+						broadcast_attempt: BroadcastAttempt {
+							broadcast_id,
+							transaction_payload: attempt.broadcast_attempt.transaction_payload,
+							threshold_signature_payload: attempt
+								.broadcast_attempt
+								.threshold_signature_payload,
+							transaction_out_id: attempt.broadcast_attempt.transaction_out_id,
+						},
+						nominee: attempt.nominee,
+					},
+				);
+			}
+		});
+
+		// Failed broadcaster storage: Vec<ValidatorId> -> BTreeSet<ValidatorId>
+		FailedBroadcasters::<T, I>::translate(|_, failed_broadcasters: Vec<T::ValidatorId>| {
```

Comment:
This shouldn't be required - BTreeSet and Vec have the same encoding.

Reply:
This translation will remove any duplicates in the vec to ensure correctness.

Reply:
Tested via unit test no migration is needed - duplication is also removed

Reply:
I've added this translate to remove any potential duplicate entries to ensure correctness.

Updated: Via unit test - without this translate(), the value obtained is correct, however the decode_len() will return incorrect result - (decode len returns length with duplicate, not the length of the set).

Though we don't expect there are dups in the old storage, I think it is worth it to add this line here to ensure the set is constructed correctly.